### PR TITLE
Make [association, aggregation] cache clearing public again

### PR DESCRIPTION
**Pull request by chrisarcand**
### Summary

This PR re-publicizes `Associations#clear_association_cache` and `Aggregations#clear_aggregation_cache`, which were privatized in https://github.com/rails/rails/pull/16989

From that PR:

> It also means that the currently undocumented clear_association_cache and clear_aggregation_cache methods are not used outside their modules and could be made private. I've done this but am ambivalent about that part, so let me know what you think.

While the preceding PR has good changes (responsibilities moved to their respective modules, limited access to the caches themselves) the privatization of the cache clearing methods serves no purpose. Clearing AR caches is a very useful thing and it seems very innocuous to have a method in the concerning module to clear out the very simple cache implemented within them.
### Other Information
- https://github.com/ManageIQ/manageiq/issues/6723
- https://github.com/ManageIQ/manageiq/pull/8668
- This partially reverts 17e0878, just bringing back the applicable tests. I did not add the API documentation mentioning this method back as it's not incredibly useful.

r? @rafaelfranca or @matthewd

### CODE REVIEW

**Comment by maclover7** [comment]
Please take a look at Code Climate, looks like there's one style guide issue with this PR. Thanks!

File: activerecord/test/cases/associations_test.rb [hunk 1]
```diff
@@ -41,6 +41,28 @@ def test_subselect
     assert_equal favs, fav2
   end
 
+  def test_clear_association_cache_stored
+    firm = Firm.find(1)
+    assert_kind_of Firm, firm
+
+    firm.send(:clear_association_cache)
```
Comment (chrisarcand) [inline activerecord/test/cases/associations_test.rb]: Whoops, that's from reverting the commit that removed this (which was after the method was privatized). Thanks.

**Comment by chrisarcand** [comment]
@rafaelfranca I had to kick the build a few times due to unrelated errors elsewhere, so this is now good to go for a glance. 🙇

**Comment by rafaelfranca** [comment]
Those methods were private API of the framework and are not supposed to be used in application. In my point of view they should also be private visibility and the change was fine since private APIs are expected to change or even be removed. That said I believe it is better for them to have private visibility.

Thank you for the pull resquest.

**Comment by Fryguy** [comment]
@rafaelfranca I understand your point about how the original change was fine since it basically went from "should have been private" to "actually private", but can you explain why you think a method for clearing caches should remain private at all?  I can't see the downside to exposing those methods.

**Comment by rafaelfranca** [comment]
Of course. The main reason I think those method should not be exposed is that users should not be aware that associations or aggregations are cached and there are ways to clean those specific caches. Cache is those cases are implementation detail of the framework and we should not leak implementation details to the public API.

There is a public API to clear all the caches in an instance and this public API is `#reload`.

**Comment by Fryguy** [comment]
Ah thanks @rafaelfranca.  The problem there is that `#reload` goes back to the database to refetch the record, and you pay a SQL penalty just for clearing a cache.  I already have the record, I just want to clear the cache in order to keep memory levels low, which is what the old `clear_association_cache` was useful for.

We originally had `#reload` to clear this cache, but we are finding that the SQL penalty for that extra query is too great.

**Comment by rafaelfranca** [comment]
Yeah, I was reviewing the linked issues, I see why you need this but I really don't think it is a good idea to Rails expose those methods. If I needed this in my application I'd be fine to be calling this method using `send` and dealing with any future removal since it is a very specific use case.

**Comment by chrisarcand** [comment]
> but I really don't think it is a good idea to Rails expose those methods

I don't see the harm and respectfully disagree (I really should have added the info about `reload` forcing a query, sorry), but that's ok :) We will continue to use the now-private API for now, I guess.

**Pull request closed without merging**